File: frontend/src/app/core/services/approval.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApprovalRequest, RequestStatus } from '../models/approval-request.model';
import { RequestType } from '../models/request-type.model';
import { CreateRequestDTO } from '../models/create-request.dto';
import { ApprovalActionDTO } from '../models/approval-action.dto';

@Injectable({
  providedIn: 'root'
})
export class ApprovalService {
  private apiUrl = `${environment.apiUrl}/requests`;

  constructor(private http: HttpClient) {}

  // POST /api/requests - Crear nueva solicitud
  createRequest(dto: CreateRequestDTO): Observable<ApprovalRequest> {
    return this.http.post<ApprovalRequest>(this.apiUrl, dto);
  }

  // GET /api/requests - Obtener todas las solicitudes
  getAllRequests(): Observable<ApprovalRequest[]> {
    return this.http.get<ApprovalRequest[]>(this.apiUrl);
  }

   // GET /api/requests/{id} - Obtener solicitud por ID
  getRequestById(id: string): Observable<ApprovalRequest> {
    return this.http.get<ApprovalRequest>(`${this.apiUrl}/${id}`);
  }

   // GET /api/requests/requester/{username} - Solicitudes del usuario
  getRequestsByRequester(username: string): Observable<ApprovalRequest[]> {
    return this.http.get<ApprovalRequest[]>(`${this.apiUrl}/requester/${username}`);
  }

   // GET /api/requests/approver/{username}/pending - Solicitudes pendientes del aprobador
  getPendingRequestsByApprover(username: string): Observable<ApprovalRequest[]> {
    return this.http.get<ApprovalRequest[]>(`${this.apiUrl}/approver/${username}/pending`);
  }

   // GET /api/requests/status/{status} - Solicitudes por estado
  getRequestsByStatus(status: RequestStatus): Observable<ApprovalRequest[]> {
    return this.http.get<ApprovalRequest[]>(`${this.apiUrl}/status/${status}`);
  }

  //  PUT /api/requests/{id}/approve - Aprobar solicitud
  approveRequest(id: string, actionDTO: ApprovalActionDTO): Observable<ApprovalRequest> {
    return this.http.put<ApprovalRequest>(`${this.apiUrl}/${id}/approve`, actionDTO);
  }

  //  PUT /api/requests/{id}/reject - Rechazar solicitud
  rejectRequest(id: string, actionDTO: ApprovalActionDTO): Observable<ApprovalRequest> {
    return this.http.put<ApprovalRequest>(`${this.apiUrl}/${id}/reject`, actionDTO);
  }

   // GET /api/requests/types - Obtener tipos de solicitud
  getAllRequestTypes(): Observable<RequestType[]> {
    return this.http.get<RequestType[]>(`${this.apiUrl}/types`);
  }

   // GET /api/requests/approver/{username}/pending/count - Contar pendientes
  countPendingRequests(username: string): Observable<{ count: number }> {
    return this.http.get<{ count: number }>(`${this.apiUrl}/approver/${username}/pending/count`);
  }
}

File: backend/src/main/java/com/bank/backend/service/ApprovalService.java
package com.bank.backend.service;

import com.bank.backend.dto.request.ApprovalActionDTO;
import com.bank.backend.dto.request.CreateApprovalRequestDTO;
import com.bank.backend.dto.response.ApprovalRequestResponseDTO;
import com.bank.backend.entity.RequestType;
import com.bank.backend.enums.RequestStatus;

import java.util.List;
import java.util.UUID;

public interface ApprovalService {

     // Crea una nueva solicitud de aprobación
    ApprovalRequestResponseDTO createRequest(CreateApprovalRequestDTO dto);

     // Obtiene una solicitud por su ID
    ApprovalRequestResponseDTO getRequestById(UUID id);

     // Obtiene todas las solicitudes
    List<ApprovalRequestResponseDTO> getAllRequests();

     // Obtiene solicitudes por solicitante
    List<ApprovalRequestResponseDTO> getRequestsByRequester(String requester);

     // Obtiene solicitudes pendientes de un aprobador
    List<ApprovalRequestResponseDTO> getPendingRequestsByApprover(String approver);

     // Obtiene solicitudes por estado
    List<ApprovalRequestResponseDTO> getRequestsByStatus(RequestStatus status);

     // Aprueba una solicitud
    ApprovalRequestResponseDTO approveRequest(UUID id, ApprovalActionDTO actionDTO);

     // Rechaza una solicitud
    ApprovalRequestResponseDTO rejectRequest(UUID id, ApprovalActionDTO actionDTO);

     // Obtiene todos los tipos de solicitud disponibles
    List<RequestType> getAllRequestTypes();

     // Cuenta solicitudes pendientes de un aprobador
    Long countPendingRequestsByApprover(String approver);
}
